pkg/component/controller/workerconfig/reconciler.go

```diff
@@ -0,0 +1,606 @@
+*/
+
+package workerconfig
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net"
+	"path/filepath"
+	"reflect"
+	"sort"
+	"strings"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/k0sproject/k0s/pkg/apis/k0s.k0sproject.io/v1beta1"
+	"github.com/k0sproject/k0s/pkg/applier"
+	"github.com/k0sproject/k0s/pkg/component"
+	"github.com/k0sproject/k0s/pkg/component/controller/leaderelector"
+	workerconfig "github.com/k0sproject/k0s/pkg/component/worker/config"
+	"github.com/k0sproject/k0s/pkg/constant"
+	kubeutil "github.com/k0sproject/k0s/pkg/kubernetes"
+
+	corev1 "k8s.io/api/core/v1"
+	rbacv1 "k8s.io/api/rbac/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/kubernetes"
+	kubeletv1beta1 "k8s.io/kubelet/config/v1beta1"
+	"k8s.io/utils/pointer"
+
+	"github.com/sirupsen/logrus"
+	"golang.org/x/exp/slices"
+	"sigs.k8s.io/yaml"
+)
+
+type resources = []*unstructured.Unstructured
+
+type mtex = sync.Mutex
+
+// Reconciler maintains ConfigMaps that hold configuration to be
+// used on k0s worker nodes, depending on their selected worker profile.
+type Reconciler struct {
+	log logrus.FieldLogger
+
+	cleaner       cleaner
+	leaderElector leaderelector.Interface
+
+	mu            mtex
+	state         interface{}
+	clusterDNSIP  net.IP
+	clusterDomain string
+}
+
+var (
+	_ component.Component  = (*Reconciler)(nil)
+	_ component.Reconciler = (*Reconciler)(nil)
+)
+
+type updateFunc = func(*snapshot) chan<- error
+
+type reconcilerStarted struct {
+	stop    func()
+	stopped <-chan struct{}
+
+	updates chan<- updateFunc
+}
+
+type reconcilerCreated struct {
+	clientFactory kubeutil.ClientFactoryInterface
+}
+
+// NewReconciler creates a new reconciler for worker configurations.
+func NewReconciler(k0sVars constant.CfgVars, nodeSpec *v1beta1.ClusterSpec, clientFactory kubeutil.ClientFactoryInterface, leaderElector leaderelector.Interface) (*Reconciler, error) {
+	log := logrus.WithFields(logrus.Fields{"component": "workerconfig.Reconciler"})
+
+	clusterDNSIPString, err := nodeSpec.Network.DNSAddress()
+	if err != nil {
+		return nil, err
+	}
+	clusterDNSIP := net.ParseIP(clusterDNSIPString)
+	if clusterDNSIP == nil {
+		return nil, fmt.Errorf("not an IP address: %q", clusterDNSIPString)
+	}
+
+	reconciler := &Reconciler{
+		log: log,
+
+		leaderElector: leaderElector,
+		clusterDomain: nodeSpec.Network.ClusterDomain,
+		clusterDNSIP:  clusterDNSIP,
+		cleaner: &kubeletConfigCleaner{
+			log: log,
+
+			dir:           filepath.Join(k0sVars.ManifestsDir, "kubelet"),
+			clientFactory: clientFactory,
+		},
+		state: reconcilerCreated{clientFactory},
+	}
+
+	return reconciler, nil
+}
+
+type reconcilerInitialized struct {
+	client kubernetes.Interface
+	apply  func(context.Context, resources) error
+}
+
+func (r *Reconciler) Init(context.Context) error {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	var created reconcilerCreated
+	switch state := r.state.(type) {
```

Comment:
Not sure, I follow, but what is the reason for this abstraction?

Reply:
`state` encapsulates the current state of the component state machine. State can be one of created, initialized, started and stopped. The Init method performs the state transition from created to initialized. The switch statement was used to assert that the Reconciler is in the right state for the transition (the "created" state).

By having an opaque state variable and one type for each possible state, one needs to use a type assertion to access the data associated with a given state. So there's no way to accidentally use something which hasn't been initialized or is destroyed already, since one needs to prove to the compiler first that the state to be accessed is actually the state we're in.

A concrete example for the Reconciler is the update channel that receives state updates. Having this in the reconcilerStarted struct instead of the Reconciler struct itself expresses that the channel is only valid as long as the Reconciler is started, not before or after. There's no way of accessing it without checking that we're in the right state beforehand, i.e. by `if state, ok := r.state.(reconcilerStarted); ok { /* safely access the update channel */ } `. At the same time, having a reference to the concrete state type can be a proof that the check has already been performed and doesn't need to be repeated, e.g. by passing on the concrete state when calling other helper functions that assume the Reconciler to be in that same state.

Returning to the code at hand: Since there's only one valid source state in which Init is allowed to be called, the switch isn't necessary. It just needs to check if the current state is "created". So I replaced this and similar checks with a simple if, e.g. like this:

```go
	if _, ok := r.state.(reconcilerCreated); !ok {
		return fmt.Errorf("cannot initialize: %T", r.state)
	}
```